core/unit_test/TestParseCmdLineArgsAndEnvVars.cpp

```diff
@@ -0,0 +1,294 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+
+#include <TestDefaultDeviceType_Category.hpp>
+
+#include <cstdlib>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+namespace Kokkos {
+namespace Impl {
+void parse_command_line_arguments(int& narg, char* arg[],
+                                  InitArguments& arguments);
+void parse_environment_variables(InitArguments& arguments);
+}  // namespace Impl
+}  // namespace Kokkos
+
+namespace {
+
+class EnvVarsHelper {
+  // do not let GTest run unit tests that set the environment concurently
+  static std::mutex mutex_;
+  std::vector<std::string> vars_;
+  // FIXME_CXX17 prefer optional
+  // store name of env var that was already set (if any)
+  // in which case unit test is skipped
+  std::unique_ptr<std::string> skip_;
+
+ public:
+  auto& skip() { return skip_; }
+  EnvVarsHelper(std::unordered_map<std::string, std::string> const& vars) {
+    mutex_.lock();
+    for (auto const& x : vars) {
+      auto const& name  = x.first;
+      auto const& value = x.second;
+      // skip unit test if env var is already set
+      if (getenv(name.c_str())) {
+        skip_ = std::make_unique<std::string>(name);
+        break;
+      }
+#ifdef _WIN32
+      int const error_code = _putenv((name + "=" + value).c_str());
+#else
+      int const error_code =
+          setenv(name.c_str(), value.c_str(), /*overwrite=*/0);
+#endif
+      if (error_code != 0) {
+        std::cerr << "failed to set environment variable '" << name << "="
+                  << value << "'\n";
+        std::abort();
+      }
+      vars_.push_back(name);
+    }
+  }
+  ~EnvVarsHelper() {
+    for (auto const& name : vars_) {
+#ifdef _WIN32
+      int const error_code = _putenv((name + "=").c_str());
+#else
+      int const error_code = unsetenv(name.c_str());
+#endif
+      if (error_code != 0) {
+        std::cerr << "failed to unset environment variable '" << name << "'\n";
+        std::abort();
+      }
+    }
+    mutex_.unlock();
+  }
+};
+std::mutex EnvVarsHelper::mutex_;
+
+class CmdLineArgsHelper {
+  int argc_;
+  std::vector<char*> argv_;
+  std::vector<std::unique_ptr<char[]>> args_;
+
+ public:
+  CmdLineArgsHelper(std::vector<std::string> const& args) : argc_(args.size()) {
```

Comment:
this could be made explicit but probably overdoing it here

Reply:
Will do in a follow-up